### PR TITLE
Prevent invalidating subscription iterator 

### PR DESCRIPTION
**Pull request by cngu**
This change shallow copies `_subscribers` before iterating over it, in-case one of the subscription listeners attempts to unsubscribe synchronously and invalidate the iterator.

### CODE REVIEW

File: src/store.js [hunk 1]
```diff
@@ -100,7 +100,7 @@ export class Store {
         handler(payload)
       })
     })
-    this._subscribers.forEach(sub => sub(mutation, this.state))
+    this._subscribers.slice().forEach(sub => sub(mutation, this.state))
```
Comment (cngu) [inline src/store.js]: I'm open to cleaner/faster ways to generate a shallow copy.

**Comment by cngu** [comment]
It can be useful in scenarios where you want to teardown a subscription when a particular mutation occurs (e.g. logout), or mutations that are guaranteed to only be committed N times during the lifetime of the app (e.g. 1 login from OAuth flow, as another login would have to reload the page).

File: src/store.js (hunk 1 above)
Comment (ktsn) [reply src/store.js]: I guess it's ok to use `slice` for that.

File: src/store.js [hunk 2]
```diff
@@ -100,7 +100,7 @@ export class Store {
         handler(payload)
       })
     })
-    this._subscribers.forEach(sub => sub(mutation, this.state))
```
Comment (ktsn) [inline src/store.js]: Could you write some comment to share the context?

Comment (ktsn) [inline src/store.js]: I think we also need to do the same thing for `_actionSubscribers` in `dispatch` method.
https://github.com/vuejs/vuex/blob/dev/src/store.js#L132

Comment (cngu) [reply src/store.js]: Done!

Comment (cngu) [reply src/store.js]: Good catch, I've shared it through a new `notifySubscribers` function.

**Comment by cngu** [comment]
Hi @ktsn, are there any other changes you'd like to see in this PR?

**Pull request merged**